File: ios/Game6561/Game/GameState.swift
import Foundation

let gridSize = 6
let tileValues = [0, 3, 9, 27, 81, 243, 729, 2187, 6561]
let winTileExp = 9
let maxHistory = 10

struct GameState: Equatable {
    var grid: [[Int]]
    var score: Int
    var best: Int
    var moves: Int
    var gameTimer: Int
    var gameWon: Bool
    var gameOver: Bool
    var combo: Int
    var totalMerges: Int
    var maxCombo: Int
    var gamesPlayed: Int
    var gamesWon: Int
    var history: [HistoryEntry]

    init(
        grid: [[Int]]? = nil,
        score: Int = 0,
        best: Int = 0,
        moves: Int = 0,
        gameTimer: Int = 0,
        gameWon: Bool = false,
        gameOver: Bool = false,
        combo: Int = 0,
        totalMerges: Int = 0,
        maxCombo: Int = 0,
        gamesPlayed: Int = 0,
        gamesWon: Int = 0,
        history: [HistoryEntry] = []
    ) {
        self.grid = grid ?? Array(repeating: Array(repeating: 0, count: gridSize), count: gridSize)
        self.score = score
        self.best = best
        self.moves = moves
        self.gameTimer = gameTimer
        self.gameWon = gameWon
        self.gameOver = gameOver
        self.combo = combo
        self.totalMerges = totalMerges
        self.maxCombo = maxCombo
        self.gamesPlayed = gamesPlayed
        self.gamesWon = gamesWon
        self.history = history
    }

    mutating func copyGrid() -> [[Int]] {
        return grid.map { $0 }
    }

    var hasEmptyCell: Bool {
        for r in 0..<gridSize {
            for c in 0..<gridSize {
                if grid[r][c] == 0 { return true }
            }
        }
        return false
    }
}

struct HistoryEntry: Equatable {
    var grid: [[Int]]
    var score: Int
    var moves: Int
    var gameWon: Bool
    var gameOver: Bool
    var combo: Int
}

struct GameStatistics: Equatable {
    var gamesPlayed: Int
    var gamesWon: Int
    var totalScore: Int
    var bestScore: Int
    var totalMoves: Int
    var maxCombo: Int
}

enum Direction: CaseIterable {
    case left, right, up, down
}

File: android/app/src/main/java/com/game6561/app/game/GameState.kt
package com.game6561.app.game

import kotlin.random.Random

const val GRID_SIZE = 6
val TILE_VALUES = listOf(0, 3, 9, 27, 81, 243, 729, 2187, 6561)
const val WIN_TILE_EXP = 9
const val MAX_HISTORY = 10

data class GameState(
    val grid: Array<IntArray> = Array(GRID_SIZE) { IntArray(GRID_SIZE) },
    val score: Int = 0,
    val best: Int = 0,
    val moves: Int = 0,
    val gameTimer: Int = 0,
    val gameWon: Boolean = false,
    val gameOver: Boolean = false,
    val combo: Int = 0,
    val totalMerges: Int = 0,
    val maxCombo: Int = 0,
    val gamesPlayed: Int = 0,
    val gamesWon: Int = 0,
    val history: List<HistoryEntry> = emptyList()
) {
    fun copyGrid(): Array<IntArray> = Array(GRID_SIZE) { r -> grid[r].copyOf() }

    fun hasEmptyCell(): Boolean {
        for (r in 0 until GRID_SIZE) {
            for (c in 0 until GRID_SIZE) {
                if (grid[r][c] == 0) return true
            }
        }
        return false
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is GameState) return false
        return score == other.score &&
                best == other.best &&
                moves == other.moves &&
                gameTimer == other.gameTimer &&
                gameWon == other.gameWon &&
                gameOver == other.gameOver &&
                combo == other.combo &&
                grid.contentDeepEquals(other.grid)
    }

    override fun hashCode(): Int {
        var result = grid.contentDeepHashCode()
        result = 31 * result + score
        result = 31 * result + best
        result = 31 * result + moves
        result = 31 * result + gameTimer
        result = 31 * result + gameWon.hashCode()
        result = 31 * result + gameOver.hashCode()
        result = 31 * result + combo
        return result
    }
}

data class HistoryEntry(
    val grid: Array<IntArray>,
    val score: Int,
    val moves: Int,
    val gameWon: Boolean,
    val gameOver: Boolean,
    val combo: Int
)

data class MergeResult(
    val newGrid: Array<IntArray>,
    val score: Int,
    val mergedCount: Int
)

data class GameStatistics(
    val gamesPlayed: Int,
    val gamesWon: Int,
    val totalScore: Int,
    val bestScore: Int,
    val totalMoves: Int,
    val maxCombo: Int
)
